### ISSUE TITLE
App irreparably frozen after accidentally queueing up a very large download

### ISSUE THREAD

**Issue by JohnnyBroccoli** [nicotine-plus/nicotine-plus#2910]
Nicotine+ version:  3.3.2

Operating System/Distribution: macOS Monterey Version 12.7.3

### Describe the bug

Nicotine+ app has been rendered completely useless via the spinning beach ball of death no matter how many times I force quit and restart it after accidentally starting up a download of an absolutely massive folder

### Expected behavior

I'd expect to be able to clear all downloads and actually use the app like normal

### Steps to reproduce the bug

The bug is omnipresent since this morning and doesn't go away regardless of how many times I force quit and restart the app

### Additional context
Screenshots, logs, stacktraces or relevant information.

I'm not sure what would be relevant or helpful, so please let me know if I can share anything that might help someone troubleshoot this issue with me (having said that, I can't do much navigating around the app as it is frozen)

**Comment by slook** [comment]
As a workaround you could edit or delete the `downloads.json` file to clear the problem downloads, it is usually located inside a hidden folder that is in somewhere like `/home/user/.local/share/nicotine/`.

**Comment by mathiascode** [comment]
Could you send your downloads.json file to my email address (on my GitHub profile)? It would be interesting to know exactly how many downloads there are.

**Comment by slook** [comment]
> Nicotine+ only freezes when trying to open the Downloads tab?

And of course that includes if the open last remembered tab on startup option is enabled and that tab is the Downloads tab, hence the application window is totally bricked, because not even switching to a different tab is possible under this circumstance.

**Comment by JohnnyBroccoli** [comment]
> Could you send your downloads.json file to my email address (on my GitHub profile)? It would be interesting to know exactly how many downloads there are. You can then remove the file, and the problem should go away.
> 
> To find the folder containing the file:
> 
>     1. In Finder, Go -> Home
> 
>     2. Press Command+Shift+. (period) to show hidden folders
> 
>     3. Go to .local/share/nicotine
> 
> 
> Also, how did you end up with so many downloads? There should've been a confirmation dialog before starting a folder download if there are too many files in it.
> 
> As far as I understand, downloads still work, and Nicotine+ only freezes when trying to open the Downloads tab?

First of all: thank you for your help. 

I have found the "download.json" file both you and slook mentioned. Unfortunately, Google wouldn't let me send it because of the size (247.5 MB); if you have any ideas of another way to get you the info, please let me know. As of now, I have made a duplicate of the file and saved it to my desktop and am planning on deleting the original file shortly after finishing this post.

I'm pretty sure it was 100,000+ files. I actually saw the warning before the download started but had a total brain fart and clicked "yes" instead of "no". I'm usually pretty good with this kind of stuff, so I felt kinda dumb about that. Then the app froze up moments later and I felt even dumber.

As to your last paragraph here: downloads do *not* still work. Nicotine+ might freeze up regardless but will 100% freeze up as soon as I try to click on the "downloads" tab.

<img width="254" alt="Screen Shot 2024-03-01 at 11 47 27 AM" src="https://github.com/nicotine-plus/nicotine-plus/assets/161789674/b4609659-5f49-426b-bc89-4fb518768857">
<img width="118" alt="Screen Shot 2024-03-01 at 11 47 47 AM" src="https://github.com/nicotine-plus/nicotine-plus/assets/161789674/14bf5b4c-26c4-456a-a006-3a42694bbda2">

**Comment by JohnnyBroccoli** [comment]
Welp....I deleted the "downloads.json" file from the hidden folder as suggested and the app still freezes up the moment I click on the downloads tab. 

Should I delete the "downloads.json.old" file too? Anything else I can try?

<img width="184" alt="Screen Shot 2024-03-01 at 12 10 03 PM" src="https://github.com/nicotine-plus/nicotine-plus/assets/161789674/204addbb-3ee2-4246-ae3f-7aa4f8c07975">

**Comment by slook** [comment]
> wouldn't let me send it because of the size (247.5 MB); if you have any ideas of another way to get you the info, please let me know

Maybe put the json file somewhere in your shares, then use the "Copy URL" function in Browse Shares to get a "slsk://..." link that you can put in the email. Alternatively, just put it somewhere obvious and tell Mat what your username is via email.

**Comment by JohnnyBroccoli** [comment]
Thank goodness. 

Deleting the "downloads.json.old" folder as well has fixed the freezing issue. If you guys would still like me to drop the "downloads.json" and/or "downloads.json.old" files in to my shared folder and send you my Nicotine+ user name via email, please let me know (especially if it'll help y'all improve this great app and/or help others avoid this issue in the future).

I really appreciate the time both of you put in to helping me with this. Very unlikely I would have figured any of this out on my own.

**Comment by JohnnyBroccoli** [comment]
On a related tip, I believe there's an option in settings to limit the amount of downloads other users can snag from you (aka uploads) at once. Maybe adding a similar option for downloads would be helpful.

Also....to the user I accidentally initiated these downloads from: sorry! I hope my blunder hasn't screwed up your ability to use the app and if it has, I hope you've found your way here for the solution.

**Comment by mathiascode** [comment]
> I have found the "download.json" file both you and slook mentioned. Unfortunately, Google wouldn't let me send it because of the size (247.5 MB); if you have any ideas of another way to get you the info, please let me know. As of now, I have made a duplicate of the file and saved it to my desktop and am planning on deleting the original file shortly after finishing this post.

That explains things. There were large performance improvements in 3.3.0, but the number of files is probably in the millions if the downloads.json file is that large.

We should do something in cases where the number of downloads is this ridiculous, but I'm not sure what. Discard downloads? What should the limit be? How is the user informed?

Feel free to send me an email containing your username, and I'll download your downloads.json file.

**Comment by JohnnyBroccoli** [comment]
> That explains things. There were large performance improvements in 3.3.0, but the number of files is probably in the millions if the downloads.json file is that large.
> 

Not quite millions. I believe it was 130,000 or so.

> We should do something in cases where the number of downloads is this ridiculous, but I'm not sure what. Discard downloads? What should the limit be? How is the user informed?
> 

Agreed. I'd say something like a 500 file queue limit based on my usage patterns (even before this, I had a limit on uploads of 200 files at a time for each user). Don't know much about the backend of applications like this but I'd think an automatic private message or notification alerting both the downloader and the uploader that they are "only allowed to have a queue of 500 files or smaller at any given time (as a safety measure to protect all users)" or something.

**Comment by slook** [comment]
It's plausible that there are use cases for initiating massive transfers, such as buddies who wish to create a mirror, for example.

**Comment by mathiascode** [comment]
> > That explains things. There were large performance improvements in 3.3.0, but the number of files is probably in the millions if the downloads.json file is that large.
> 
> Not quite millions. I believe it was 130,000 or so.

Looks like it's 1.3 million downloads. Impressive!

**Comment by JohnnyBroccoli** [comment]
LOL guess I also misread the warning prompt before accidentally clicking on "yes" instead of "no"

**Comment by slook** [comment]
Perhaps it would help if the dialog button labels were changed to something like "Cancel" and "Download".

**Comment by mathiascode** [comment]
> Perhaps it would help if the dialog button labels were changed to something like "Cancel" and "Download".

Seems like a good idea.

Looks like another case where almost all time is spent adding rows to the treeview, because Gtk.TreeView insertions are quadratic (O(n^2)).

**Comment by JohnnyBroccoli** [comment]
> Perhaps it would help if the dialog button labels were changed to something like "Cancel" and "Download".

Sounds like a good idea.

Maybe also add a little note saying something like "proceed with caution when attempting to download more than x amount of files at a time, as downloads of this size have been known to brick Nicotine+".

**Comment by slook** [comment]
> We should do something in cases where the number of downloads is this ridiculous, but I'm not sure what.

What about implementing Threading and/or Multiprocessing during any potentially long (of more than, say, 100 files) user-requested operations, such as while loading transfers history file and while batch-adding transfers from a multi-selection. The Tree View container can be hidden while it is busy and replaced with a label or progress bar widget.

> What should the limit be? 

Ideally there should not be any kind of artificial limit imposed by the user interface, because in a perfect world that should be only limited by the capabilities of the system and network environment. If peers want to exchange millions of files between themselves, then so be it (provided that is truly want they actually intended to do).

> How is the user informed?

Preferably, display a Progress Bar popup or widget in the Status Bar (similar to the "Scanning Shares" spinner) along with a Cancel [X] button to raise some kind of flag to let the core know that the user wants to abort the requested operation. It would be a bonus if the status spinner label could indicate "Added X of X items" or such like, unless it is indeterminate then "Added X items".

If the status bar is frozen and cannot be clicked on, then a progress window could popup after a certain number of seconds have elapsed (say, 10 seconds) since the beginning of the operation. A button to cancel the current operation could be provided, or otherwise the window need only contain a plain label. A progress bar widget and other decorations could be added later.

> Discard downloads? 

Such kind of "is_user_wanting_to_abort" flag can then be caught inside the calling core loop to break it. After the loop is broken in this way then any unprocessed downloads would, in effect, be discarded.

**Comment by DeathStalker77** [comment]
Yeah, I would say any single download request that exceeds 500k files should be prevented, IF the program cannot handle it properly.  _Definitely_ over 1 million!  Even MY queue has never been that large! :-D 

I frequently go well over 150,000 files queued, so I do not it any way believe the limit should be that low.  WARNINGS on over 1,000 is fine - maybe adjust the wording/color/format of WARNINGS for larger single requests.

--- DS

**Comment by mathiascode** [comment]
Just wanted to let you know that there have been some large improvements related to populating trees/lists. Loading the shares of a user with 2 million files and 175k folders, for example:
- 13 seconds to populate the GUI in Nicotine+ 3.3.4
- 4 seconds in Nicotine+ 3.3.5

And the huge `downloads.json` file actually loads now!

![downloads](https://github.com/user-attachments/assets/9a79174f-6425-48cb-abfe-709fc6ae380f)

Small improvements can probably still be made to the loading times of the transfer list files.

To put into perspective how far we've come: 4 years ago, when I started contributing to Nicotine+, it would hang with only ~2k rows in the download list.

**Comment by mathiascode** [comment]
I'd say this is good enough now.
